Sources/Algorithms/Compacted.swift

```diff
@@ -0,0 +1,195 @@
+  where Base.Element == Element? {
+
+  @usableFromInline
+  let base: Base
+  
+  @inlinable
+  init(base: Base) {
+    self.base = base
+    let idx = base.firstIndex(where: { $0 != nil }) ?? base.endIndex
+    self.startIndex = Index(base: idx)
+  }
+  
+  public struct Index {
+    @usableFromInline
+    let base: Base.Index
+    
+    @inlinable
+    init(base: Base.Index) {
+      self.base = base
+    }
+  }
+  
+  public var startIndex: Index
+  
+  @inlinable
+  public var endIndex: Index { Index(base: base.endIndex) }
+  
+  @inlinable
+  public subscript(position: Index) -> Element {
+    base[position.base]!
+  }
+  
+  @inlinable
+  public func index(after i: Index) -> Index {
+    guard i != endIndex else {
+      fatalError("Index out of bounds")
+    }
+    let baseIdx = base.index(after: i.base)
+    guard let idx = base[baseIdx...].firstIndex(where: { $0 != nil })
+      else { return endIndex }
+    return Index(base: idx)
+  }
+}
+
+extension CompactedCollection: BidirectionalCollection
+  where Base: BidirectionalCollection {
+
+  @inlinable
+  public func index(before i: Index) -> Index {
+    guard i != startIndex else {
+      fatalError("Index out of bounds")
+    }
+    
+    let baseIdx = base.index(before: i.base)
+    guard let idx = base[...baseIdx].lastIndex(where: { $0 != nil })
+      else { return Index(base: base.startIndex) }
+    return Index(base: idx)
+  }
+}
+
+extension CompactedCollection.Index: Comparable {  
+  @inlinable
+  public static func < (lhs: CompactedCollection.Index,
+                        rhs: CompactedCollection.Index) -> Bool {
+    lhs.base < rhs.base
+  }
+}
+
+extension CompactedCollection.Index: Hashable
+  where Base.Index: Hashable {}
+
+extension Collection {
+  /// Complexity: O(*n*) where *n*  is the number of elements in the
+  /// base collection.
+  @inlinable
+  public func compacted<Unwrapped>() -> CompactedCollection<Self, Unwrapped>
+    where Element == Unwrapped? {
+    CompactedCollection(base: self)
+  }
+}
+
+extension CompactedSequence: LazySequenceProtocol
+  where Base: LazySequenceProtocol {}
+
+extension CompactedCollection: RandomAccessCollection
+  where Base: RandomAccessCollection {}
+extension CompactedCollection: LazySequenceProtocol
+  where Base: LazySequenceProtocol {}
+extension CompactedCollection: LazyCollectionProtocol
+  where Base: LazyCollectionProtocol {}
+
+
+// Hashable and Equatable conformance are based on each non-nil
+// element on base collection.
+extension CompactedSequence: Equatable
+  where Base: Equatable, Base.Element: Equatable {
+
+  @inlinable
+  public static func ==(lhs: CompactedSequence,
+                        rhs: CompactedSequence) -> Bool {
+    lhs.compactMap({ $0 })
+      .elementsEqual(rhs.compactMap({ $0 }))
+  }
+}
+
+extension CompactedSequence: Hashable
+  where Base: Hashable, Base.Element: Hashable {
+  @inlinable
+  public func hash(into hasher: inout Hasher) {
+    for element in base {
```

Comment:
This could just be `for element in self`, so that you don't need to re-unwrap. (Ditto with the collection version)